bigtable/client/row_reader.h

```diff
@@ -0,0 +1,173 @@
+  /// Signifies that there is no limit on the number of rows to read.
+  static int constexpr NO_ROWS_LIMIT = 0;
+
+  RowReader(std::shared_ptr<DataClient> client, absl::string_view table_name,
+            RowSet row_set, int rows_limit, Filter filter,
+            std::unique_ptr<RPCRetryPolicy> retry_policy,
+            std::unique_ptr<RPCBackoffPolicy> backoff_policy);
+
+  using iterator = RowReaderIterator;
+
+  /**
+   * Input iterator over rows in the response.
+   *
+   * The returned iterator is a single-pass input iterator that reads
+   * rows from the RowReader when incremented. The first row may be
+   * read when the iterator is constructed.
+   *
+   * Creating, and particularly incrementing, multiple iterators on
+   * the same RowReader is unsupported and can produce incorrect
+   * results.
+   *
+   * Retry and backoff policies are honored.
+   *
+   * @throws std::runtime_error if the read failed after retries.
+   */
+  iterator begin();
+
+  /// End iterator over the rows in the response.
+  iterator end();
+
+  /**
+   * Stop the read call and clean up the connection.
+   *
+   * Invalidates iterators.
+   */
+  void Cancel();
+
+ private:
+  /**
+   * Read and parse the next row in the response.
+   *
+   * @param row receives the next row on success, and is reset on failure or if
+   * there are no more rows.
+   *
+   * This call possibly blocks waiting for data until a full row is available.
+   */
+  void Advance(absl::optional<Row>& row);
+
+  /// Called by Advance(), does not handle retries.
+  grpc::Status AdvanceOrFail(absl::optional<Row>& row);
+
+  /**
+   * Move the index to the next chunk, reading data if needed.
+   *
+   * Returns false if no more chunks are available.
+   */
+  bool NextChunk();
+
+  /// Sends the ReadRows request to the stub.
+  void MakeRequest();
+
+  /// The input iterator returned by begin() and end()
+  class RowReaderIterator : public std::iterator<std::input_iterator_tag, Row> {
+   public:
+    RowReaderIterator(RowReader* owner, bool is_end) : owner_(owner), row_() {}
+
+    RowReaderIterator& operator++();
+    RowReaderIterator operator++(int) {
+      RowReaderIterator tmp(*this);
+      operator++();
+      return tmp;
+    }
+
+    constexpr const Row* operator->() const { return row_.operator->(); }
+    Row* operator->() { return row_.operator->(); }
+
+    constexpr const Row& operator*() const & { return row_.operator*(); }
+    Row& operator*() & { return row_.operator*(); }
+    constexpr const Row&& operator*() const && {
+      return absl::move(row_.operator*());
+    }
+    Row&& operator*() && { return std::move(row_.operator*()); }
+
+    bool operator==(RowReaderIterator const& that) const {
+      // All non-end iterators are equal.
+      return bool(row_) == bool(that.row_);
+    }
+
+    bool operator!=(RowReaderIterator const& that) const {
+      return !(*this == that);
+    }
+
+   private:
+    RowReader* owner_;
+    absl::optional<Row> row_;
```

Comment:
I have been thinking for `std::istream_iterator` for inspiration on this.  Your implementation seems like a modern version of the `libstdc++` one:

https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-3.4/classstd_1_1istream__iterator.html

I think that is pretty cool.